Fix assigning tensor with requires_grad as constant in export

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137997

When we insert cojstants into unlifted graph, we need to detach them if they require grad BUT when we detach we need to preserve the original aliasing information.

Differential Revision: [D64406859](https://our.internmc.facebook.com/intern/diff/D64406859/)

tugsbayasgalan committed:
Fix assigning tensor with requires_grad as constant in export

When we insert cojstants into unlifted graph, we need to detach them if they require grad.

Differential Revision: [D64406859](https://our.internmc.facebook.com/intern/diff/D64406859/)

[ghstack-poisoned]

avikchaudhuri reviewed (changes requested):
Test please

tugsbayasgalan committed:
Update on "Fix assigning tensor with requires_grad as constant in export"


When we insert cojstants into unlifted graph, we need to detach them if they require grad.

Differential Revision: [D64406859](https://our.internmc.facebook.com/intern/diff/D64406859/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix assigning tensor with requires_grad as constant in export"


When we insert cojstants into unlifted graph, we need to detach them if they require grad.

Differential Revision: [D64406859](https://our.internmc.facebook.com/intern/diff/D64406859/)

[ghstack-poisoned]

avikchaudhuri commented on torch/export/_unlift.py:
```diff
@@ -309,6 +318,19 @@ def _create_stateful_graph_module(
             if isinstance(value, torch.Tensor):
+                if value.requires_grad:
+                    warnings.warn(
+                        f"A model attribute `{const_name}` requires gradient. "
```
no ending period

tugsbayasgalan replied:
sg

avikchaudhuri commented on torch/export/_unlift.py:
```diff
@@ -309,6 +318,19 @@ def _create_stateful_graph_module(
+                    if value in original_tensor_to_detached_tensor:
+                        value = original_tensor_to_detached_tensor[value]
+                    else:
+                        detached_value = value.detach()
```
maybe this code should be refactored into a util?

tugsbayasgalan replied:
It is bit hard to make it a util because this is only relevant in this specific location.

avikchaudhuri reviewed (approved):
Rebase and check CI before landing...

tugsbayasgalan committed:
Update on "Fix assigning tensor with requires_grad as constant in export"


When we insert cojstants into unlifted graph, we need to detach them if they require grad BUT when we detach we need to preserve the original aliasing information.

Differential Revision: [D64406859](https://our.internmc.facebook.com/intern/diff/D64406859/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix assigning tensor with requires_grad as constant in export"


When we insert cojstants into unlifted graph, we need to detach them if they require grad BUT when we detach we need to preserve the original aliasing information.

Differential Revision: [D64406859](https://our.internmc.facebook.com/intern/diff/D64406859/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64406859).
